pkg/monitors/nagios/nagios.go

```diff
@@ -0,0 +1,253 @@
+	// Enforce interval greater than command timeout
+	if conf.IntervalSeconds < conf.Timeout {
+		return fmt.Errorf("configured timeout must be lower than intervalSeconds")
+	}
+	// Init cache used to avoid sending duplicate event for each interval
+	c := cache.New(1*time.Minute, 1*time.Hour)
+
+	// Start the metric gathering process here
+	var ctx context.Context
+	ctx, m.cancel = context.WithCancel(context.Background())
+	utils.RunOnInterval(ctx, func() {
+		// Run command
+		stdout, stderr, err := runCommand(conf.Command, conf.Timeout, conf.IntervalSeconds)
+		state, err := getExitCode(err, stdout)
+
+		// Send datapoint with only state
+		m.Output.SendDatapoints([]*datapoint.Datapoint{
+			datapoint.New(
+				nagiosState,
+				dimensions,
+				datapoint.NewIntValue(int64(state)),
+				datapoint.Gauge,
+				time.Time{}),
+		}...)
+
+		properties, diffProperties := makeProperties(state, err, stdout, stderr)
+		m.logger.Warn(properties)
+		m.logger.Warn(diffProperties)
+		// Compare with previous event if it exists
+		sendEvent := true
+		if x, found := c.Get(cacheKey); found {
+			lastProperties := x.(map[string]interface{})
+			if diff := deep.Equal(diffProperties, lastProperties); diff == nil {
+				m.logger.Debug("a same avent has already been send, do not send again")
+				sendEvent = false
+			}
+		}
+
+		// Do not send duplicate event
+		if sendEvent {
+			// Send event with command context
+			m.Output.SendEvent(
+				event.NewWithProperties(
+					nagiosState,
+					event.AGENT,
+					dimensions,
+					properties,
+					time.Time{}),
+			)
+			// update event properties in cache
+			c.Set(cacheKey, diffProperties, cache.NoExpiration)
+		}
+
+	}, time.Duration(conf.IntervalSeconds)*time.Second)
+
+	return nil
+}
+
+// Shutdown the monitor
+func (m *Monitor) Shutdown() {
+	// Stop any long-running go routines here
+	if m.cancel != nil {
+		m.cancel()
+	}
+}
+
+func runCommand(command string, termTimeout int, killTimeout int) (stdout []byte, stderr []byte, err error) {
+	var cmdOut, cmdErr bytes.Buffer
+	var killTimer *time.Timer
+
+	// Parse command string with args
+	splitCmd, err := shellquote.Split(command)
+	if err != nil || len(splitCmd) == 0 {
+		return nil, nil, fmt.Errorf("exec: unable to parse command, %s", err)
+	}
+	// Prepare command exec
+	cmd := exec.Command(splitCmd[0], splitCmd[1:]...)
+	cmd.Stdout = &cmdOut
+	cmd.Stderr = &cmdErr
+	// Start command
+	if err = cmd.Start(); err != nil {
+		return
+	}
+
+	// After user timeout, try to gracefully stop the process with sigterm
+	termTimer := time.AfterFunc(time.Duration(termTimeout)*time.Second, func() {
+		err = cmd.Process.Signal(syscall.SIGTERM)
+		if err != nil {
+			return
+		}
+		// Kill the process to respect the monitor intervalSeconds
+		killTimer = time.AfterFunc(time.Duration(killTimeout)*time.Second, func() {
+			err = cmd.Process.Signal(syscall.SIGTERM)
+			if err != nil {
+				return
+			}
+		})
+	})
+	err = cmd.Wait()
+	// Shutdown timers
+	if killTimer != nil {
```

Comment:
This doesn't seem thread-safe since killTimer is created in a  separate goroutine.

Reply:
is it better now ? else please could you advice ? thanks